### PR TITLE
[Maintenance] Bump plugin to Sylius 2.0

### PR DESCRIPTION
**Pull request by mpysiak**
| Q               | A
| --------------- | -----
| Branch?         | 2.0
| Bug fix?        | no
| New feature?    | no
| Related tickets | n/a

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -27,51 +27,31 @@ jobs:
         strategy:
             fail-fast: false
             matrix:
-                php: ["8.3"]
-                symfony: ["^5.4.21", "^6.4"]
-                sylius: ["~1.13.0", "~1.14.0"]
-                mysql: ["8.4"]
+                php: ["8.2", "8.3"]
+                symfony: ["^6.4", "^7.1"]
+                sylius: ["2.0.0"]
```
Comment (diimpp) [inline .github/workflows/build.yml]: Is it necessary to specify patch version as I think build will be stuck at 2.0.0 even after 2.0.x release?

File: composer.json [hunk 2]
```diff
@@ -19,19 +19,20 @@
...
-        "sylius/sylius": "~1.13.0 || ~1.14.0",
-        "symfony/messenger": "^5.4.21 || ^6.4"
+        "php": "^8.2",
+        "knplabs/knp-snappy-bundle": "^1.10",
+        "myclabs/php-enum": "^1.8",
+        "sylius/resource-bundle": "^1.12",
+        "sylius/sylius": "^2.0",
+        "symfony/messenger": "^6.4 || ^7.1",
+        "php-http/discovery": "^1.20",
+        "winzou/state-machine": "^0.4",
```
Comment (diimpp) [inline composer.json]: Shouldn't plugin be upgraded to symfony/workflow for 2.0?

File: src/Creator/RefundCreator.php [hunk 3]
```diff
@@ -22,14 +22,14 @@
...
     /** @param OrderRepositoryInterface<OrderInterface> $orderRepository */
     public function __construct(
-        private readonly RefundFactoryInterface $refundFactory,
-        private readonly RemainingTotalProviderInterface $remainingTotalProvider,
-        private readonly OrderRepositoryInterface $orderRepository,
-        private readonly ObjectManager $refundManager,
+        private RefundFactoryInterface $refundFactory,
+        private RemainingTotalProviderInterface $remainingTotalProvider,
+        private OrderRepositoryInterface $orderRepository,
+        private ObjectManager $refundManager,
```
Comment (diimpp) [inline src/Creator/RefundCreator.php:32]: I think it makes sense to replace every `ObjectManager` with `EntityManagerInterface` since odm support is dropped. (There are 12 instances of `ObjectManager` in the codebase.)

```suggestion
        private EntityManagerInterface $refundManager,
```

File: src/templates/order/admin/content/header/title_block/actions/refunds.html.twig [hunk 4]
```diff
@@ -0,0 +1,8 @@
+{% set order = hookable_metadata.context.resource %}
+
+    {% set path = path('sylius_refund_order_refunds_list', {'orderNumber': order.number,}) %}
+    <a class="dropdown-item" href="{{ path }}" {{ sylius_test_html_attribute('refunds') }}>
+        {{ ux_icon('tabler:repeat', {'class': 'icon dropdown-item-icon'}) }}
+        {{ 'sylius_refund.ui.refunds'|trans }}
+    </a>
+
```
Comment (diimpp) [inline src/templates/order/admin/content/header/title_block/actions/refunds.html.twig]:
```suggestion
```

File: src/templates/order/admin/content/sections/payments/item/name.html.twig [hunk 5]
```diff
@@ -0,0 +1,5 @@
+{% set refund_payment = hookable_metadata.context.refund_payment %}
+
+<td class="py-2">
+    <strong {{ sylius_test_html_attribute('refundd-payment-method') }}>{{ refund_payment.paymentMethod }}</strong>
```
Comment (diimpp) [inline src/templates/order/admin/content/sections/payments/item/name.html.twig]:
```suggestion
    <strong {{ sylius_test_html_attribute('refund-payment-method') }}>{{ refund_payment.paymentMethod }}</strong>
```

File: src/templates/order/admin/refund/content/header/title_block/actions.html.twig [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+                {{ 'sylius.ui.actions'|trans }}
+            </button>
+            <div class="dropdown-menu dropdown-menu-end">
+                {% hook 'actions' with { order } %}
+            </div>
+        </div>
+    </div>
+</div>
+
+
```
Comment (diimpp) [inline src/templates/order/admin/refund/content/header/title_block/actions.html.twig]:
```suggestion
```

File: src/templates/order/admin/refund/content/header/title_block/actions/refund_all.html.twig [hunk 7]
```diff
@@ -0,0 +1,5 @@
+<button type="submit" class="dropdown-item" {{ sylius_test_html_attribute('clear-all') }}>
+    {{ ux_icon('tabler:repeat', {'class': 'icon dropdown-item-icon'}) }}
+    {{ 'sylius_refund.ui.refund_all'|trans }}
+</button>
+
```
Comment (diimpp) [inline src/templates/order/admin/refund/content/header/title_block/actions/refund_all.html.twig]:
```suggestion
```

File: src/templates/order/admin/creditMemos.html.twig
Comment (diimpp) [inline src/templates/order/admin/creditMemos.html.twig:1]: Do you plan to rename all templates to snake_case? This one still at camelCase.

File: tests/Application/assets/admin/entry.js
Comment (diimpp) [inline tests/Application/assets/admin/entry.js:1]: I think this file still exists, but now empty. But I'm not sure if GH shows this right.

If true, then this file as well `tests/Application/assets/shop/entry.js`

File: composer.json [hunk 8]
```diff
@@ -19,19 +19,20 @@
...
-        "php": "^8.1",
-        "knplabs/knp-snappy-bundle": "^1.7",
-        "myclabs/php-enum": "^1.7",
-        "php-http/message-factory": "^1.1",
-        "sylius/resource-bundle": "^1.9",
-        "sylius/sylius": "~1.13.0 || ~1.14.0",
-        "symfony/messenger": "^5.4.21 || ^6.4"
+        "php": "^8.2",
+        "knplabs/knp-snappy-bundle": "^1.10",
+        "myclabs/php-enum": "^1.8",
```
Comment (diimpp) [inline composer.json:24]: I would say it's high time to drop pseudo enum support and use sylius/sylius style consts (or php enums)

**Comment by mpysiak** [comment]
Thanks @diimpp for your code review! However, this PR is still very much a work in progress, and I didn't mark it as a draft :)

**Comment by diimpp** [comment]
> Thanks @diimpp for your code review! However, this PR is still very much a work in progress, and I didn't mark it as a draft :)

@mpysiak You're welcome. Please ping me when it's done. :)

Do you plan to refactor how refund items entities are structured or it's more of the maintenance PR? There is this issue from 2020 https://github.com/Sylius/RefundPlugin/issues/193, which I think still actual. Described problem is more about letting people develop based of this plugin, rather then solving immediate bug.

tldr: Add "root" entity `Refund` (Or `RefundRequest`) to make relationship between `RefundUnit`, `RefundPayment` and `CreditMemo`.

File: .github/workflows/build.yaml [hunk 9]
```diff
@@ -27,45 +27,46 @@ jobs:
...
             matrix:
-                php: ["8.3"]
-                symfony: ["^5.4.21", "^6.4"]
-                sylius: ["~1.13.0", "~1.14.0"]
-                database: ["mysql", "postgres"]
-                mysql: ["8.4"]
-                postgres: ["15.8"]
+                php: ["8.2", "8.3"]
+                symfony: ["^6.4", "^7.1"]
+                sylius: ["~2.0"]
```
Comment (GSadee) [inline .github/workflows/build.yaml:32]: I would be in favour of limiting it to just one minor, as we did before
```suggestion
                sylius: ["~2.0.0"]
```

File: .github/workflows/build.yaml [hunk 10]
```diff
@@ -27,45 +27,46 @@ jobs:
...
                 node: ["20.x"]
+                mysql: ["8.4"]
+                database: ["mysql"]
```
Comment (GSadee) [inline .github/workflows/build.yaml:35]:
```suggestion
                database: ["mysql"]
                mysql: ["8.4"]
```

File: composer.json [hunk 11]
```diff
@@ -19,19 +19,18 @@
...
+        "sylius/resource-bundle": "^1.12",
+        "sylius/sylius": "^2.0",
```
Comment (GSadee) [inline composer.json:26]:
```suggestion
        "sylius/sylius": "~2.0.0",
```

File: composer.json (hunk 8 above)
Comment (mpysiak) [reply composer.json:24]: We would like to do that. There is another task for that. This will be done in another PR

File: composer.json [hunk 12]
```diff
@@ -40,21 +39,20 @@
...
-        "symfony/dotenv": "^5.4.21 || ^6.4",
-        "symfony/intl": "^5.4 || ^6.4",
-        "symfony/web-profiler-bundle": "^5.4.21 || ^6.4",
-        "symfony/webpack-encore-bundle": "^1.15"
+        "symfony/browser-kit": "^6.4 || ^7.1",
+        "symfony/debug-bundle": "^6.4 || ^7.1",
+        "symfony/dotenv": "^6.4 || ^7.1",
+        "symfony/intl": "^6.4 || ^7.1",
+        "symfony/web-profiler-bundle": "^6.4 || ^7.1",
+        "symfony/webpack-encore-bundle": "^2.2.0"
```
Comment (GSadee) [inline composer.json:55]:
```suggestion
        "symfony/webpack-encore-bundle": "^2.2"
```

File: composer.json [hunk 13]
```diff
@@ -40,21 +39,20 @@
...
         "friends-of-behat/variadic-extension": "^1.3",
-        "matthiasnoback/symfony-config-test": "^4.2",
-        "matthiasnoback/symfony-dependency-injection-test": "^4.2",
+        "matthiasnoback/symfony-config-test": "^5.1",
+        "matthiasnoback/symfony-dependency-injection-test": "^5.1",
         "phpspec/phpspec": "^7.2",
         "phpstan/phpstan": "^1.6",
         "phpstan/phpstan-webmozart-assert": "^1.1",
         "phpunit/phpunit": "^9.5",
         "polishsymfonycommunity/symfony-mocker-container": "^1.0",
```
Comment (GSadee) [inline composer.json:48]: Do we need this package? 🤔

**Comment by mpysiak** [comment]
@diimpp This PR is mainly Maintenance - just to bump to Sylius 2.0. I've removed deprecations, add some small refactors. But PR is getting huge, so it will be probably merged (after CR from the team). Comments and fixes will be done in other commits, to make it small and more transparent 🖖

File: src/Resources/config/app/config.yml [hunk 14]
```diff
@@ -19,19 +19,33 @@ sylius_mailer:
...
+    messenger:
+        buses:
+            sylius.command_bus: &command_bus
+                middleware:
+                    - 'validation'
+                    - 'doctrine_transaction'
+            sylius.event_bus: &event_bus
+                default_middleware: allow_no_handlers
+            sylius_refund_plugin.command_bus: *command_bus
+            sylius_refund_plugin.event_bus: *event_bus
```
Comment (GSadee) [inline src/Resources/config/app/config.yml:32]: I guess, this config could be removed and we can only use Sylius buses 🤔

File: src/Resources/config/app/events.yaml
Comment (GSadee) [inline src/Resources/config/app/events.yaml:1]: Remove this file entirely

File: src/Resources/config/integrations/winzou_state_machine.yaml
Comment (GSadee) [inline src/Resources/config/integrations/winzou_state_machine.yaml:1]: Missing blank line at the end of file

File: src/templates/credit_memos/content/sections/table.html.twig
Comment (GSadee) [inline src/templates/credit_memos/content/sections/table.html.twig:1]: This file looks like it would be a good idea to split it using hooks in case someone needs to add new column for example

File: src/templates/credit_memos/details.html.twig
Comment (GSadee) [inline src/templates/credit_memos/details.html.twig:1]: This file seems not to be needed anymore if I see correctly

File: src/templates/credit_memos/list.html.twig
Comment (GSadee) [inline src/templates/credit_memos/list.html.twig:1]: I would be nice to put it in the `templates/admin` directory 🤔

File: src/templates/order/admin/content/header/title_block/actions/refunds.html.twig [hunk 15]
```diff
@@ -0,0 +1,7 @@
+{% set order = hookable_metadata.context.resource %}
+
+    {% set path = path('sylius_refund_order_refunds_list', {'orderNumber': order.number,}) %}
+    <a class="dropdown-item" href="{{ path }}" {{ sylius_test_html_attribute('refunds') }}>
+        {{ ux_icon('tabler:repeat', {'class': 'icon dropdown-item-icon'}) }}
+        {{ 'sylius_refund.ui.refunds'|trans }}
+    </a>
```
Comment (GSadee) [inline src/templates/order/admin/content/header/title_block/actions/refunds.html.twig:7]: Not needed indentation

File: src/templates/order/admin/refund/content/header/title_block/actions/clear_all.html.twig [hunk 16]
```diff
@@ -0,0 +1,7 @@
+{% set disable_button = hookable_metadata.context.disable_button %}
+
+<button type="submit" class="dropdown-item" {{ sylius_test_html_attribute('clear-all') }} {{ disable_button }}>
+    {{ ux_icon('tabler:circle-check', {'class': 'icon dropdown-item-icon'}) }}
+    {{ 'sylius_refund.ui.clear_refunds'|trans }}
+</button>
+
```
Comment (GSadee) [inline src/templates/order/admin/refund/content/header/title_block/actions/clear_all.html.twig:7]:
```suggestion
<button type="submit" class="dropdown-item" {{ sylius_test_html_attribute('clear-all') }} {{ disable_button }}>
    {{ ux_icon('tabler:circle-check', {'class': 'icon dropdown-item-icon'}) }}
    {{ 'sylius_refund.ui.clear_refunds'|trans }}
</button>
```

File: src/templates/order/admin/refund/content/sections/form/table.html.twig [hunk 17]
```diff
@@ -0,0 +1,10 @@
+<table id="refunds" class="ui compact celled table">
+    <thead>
+    <tr>
+        {% hook 'table.header' %}
+    </tr>
+    </thead>
+    <tbody>
+        {% hook 'table.body' %}
+    </tbody>
```
Comment (GSadee) [inline src/templates/order/admin/refund/content/sections/form/table.html.twig:9]: It could be confusing not to have a `table` hook, I don't know if it wouldn't be better to add an additional template with it and trigger these two there 🤔

File: src/templates/order/admin/refund/content/sections/form/table/body/item/refund_button.html.twig [hunk 18]
```diff
@@ -0,0 +1,14 @@
+{% set unit = hookable_metadata.context.unit %}
+
+<td class="aligned collapsing">
+    <button
+            data-refund="{{ unit_refund_left(unit.id, constant('Sylius\\RefundPlugin\\Model\\RefundType::ORDER_ITEM_UNIT'), unit.total) }}"
+            type="button"
+            class="btn"
+            {% if not can_unit_be_refunded(unit.id, constant('Sylius\\RefundPlugin\\Model\\RefundType::ORDER_ITEM_UNIT')) %}disabled{% endif %}
+            {{ sylius_test_html_attribute('refund-item-button') }}
```
Comment (GSadee) [inline src/templates/order/admin/refund/content/sections/form/table/body/item/refund_button.html.twig:9]: Wrong indentation

File: src/templates/order/admin/refund/content/sections/form/table/shipment.html.twig [hunk 19]
```diff
@@ -1,5 +1,6 @@
-{% import '@SyliusAdmin/Common/Macro/money.html.twig' as money %}
+{% import "@SyliusCore/Common/Macro/money.html.twig" as money %}
 
+{% set order = hookable_metadata.context.order %}
 {% set shipmentRefundType = constant('Sylius\\RefundPlugin\\Model\\RefundType::SHIPMENT') %}
```
Comment (GSadee) [inline src/templates/order/admin/refund/content/sections/form/table/shipment.html.twig:4]: Use snake_case notation

File: src/templates/order/shop/_creditMemos.html.twig
Comment (GSadee) [inline src/templates/order/shop/_creditMemos.html.twig:1]: This one should be refactored

File: tests/Behat/Context/Application/RefundingContext.php [hunk 20]
```diff
@@ -25,14 +25,17 @@
...
-        private OrderRepositoryInterface $orderRepository,
-        private RepositoryInterface $refundRepository,
-        private RemainingTotalProviderInterface $remainingTotalProvider,
-        private MessageBusInterface $commandBus,
-        private BehatEmailCheckerInterface|CoreEmailCheckerInterface $emailChecker
+        private readonly OrderRepositoryInterface $orderRepository,
+        private readonly RepositoryInterface $refundRepository,
+        private readonly RemainingTotalProviderInterface $remainingTotalProvider,
+        private readonly MessageBusInterface $commandBus,
+        private readonly BehatEmailCheckerInterface $emailChecker
```
Comment (GSadee) [inline tests/Behat/Context/Application/RefundingContext.php:38]: The entire class could be readonly, the same applies to below classes

**Review by GSadee — APPROVED** [review]
In the separate PRs:

- add an UPGRADE file with all these changes
- change the structure to the new recommended one by moving directories from `src/Resources` to main directory

File: src/Creator/RefundCreator.php (hunk 3 above)
Comment (mpysiak) [reply src/Creator/RefundCreator.php:32]: @diimpp done in #452
